internal/project/repository/eventsourcing/model.go

```diff
@@ -612,3 +756,130 @@ func (o *OIDCConfig) setData(event *es_models.Event) error {
 	}
 	return nil
 }
+
+func (p *Project) appendAddGrantEvent(event *es_models.Event) error {
+	grant := new(ProjectGrant)
+	err := grant.getData(event)
+	if err != nil {
+		return err
+	}
+	grant.ObjectRoot.CreationDate = event.CreationDate
+	p.Grants = append(p.Grants, grant)
+	return nil
+}
+
+func (p *Project) appendChangeGrantEvent(event *es_models.Event) error {
+	grant := new(ProjectGrant)
+	err := grant.getData(event)
+	if err != nil {
+		return err
+	}
+	for i, g := range p.Grants {
+		if g.GrantID == grant.GrantID {
+			p.Grants[i].getData(event)
+		}
+	}
+	return nil
+}
+
+func (p *Project) appendGrantStateEvent(event *es_models.Event, state model.ProjectGrantState) error {
+	grant := new(ProjectGrant)
+	err := grant.getData(event)
+	if err != nil {
+		return err
+	}
+	for i, g := range p.Grants {
+		if g.GrantID == grant.GrantID {
+			g.State = int32(state)
+			p.Grants[i] = g
+		}
+	}
+	return nil
+}
+
+func (p *Project) appendRemoveGrantEvent(event *es_models.Event) error {
+	grant := new(ProjectGrant)
+	err := grant.getData(event)
+	if err != nil {
+		return err
+	}
+	for i, g := range p.Grants {
+		if g.GrantID == grant.GrantID {
+			p.Grants[i] = p.Grants[len(p.Grants)-1]
+			p.Grants[len(p.Grants)-1] = nil
+			p.Grants = p.Grants[:len(p.Grants)-1]
+		}
+	}
+	return nil
+}
+
+func (g *ProjectGrant) getData(event *es_models.Event) error {
+	g.ObjectRoot.AppendEvent(event)
+	if err := json.Unmarshal(event.Data, g); err != nil {
+		logging.Log("EVEN-4h6gd").WithError(err).Error("could not unmarshal event data")
+		return err
+	}
+	return nil
+}
+
+func (p *Project) appendAddGrantMemberEvent(event *es_models.Event) error {
+	member := &ProjectGrantMember{}
+	err := member.getData(event)
+	if err != nil {
+		return err
+	}
+	member.ObjectRoot.CreationDate = event.CreationDate
+	for _, g := range p.Grants {
+		if g.GrantID == member.GrantID {
+			g.Members = append(g.Members, member)
+		}
+	}
+	return nil
+}
+
+func (p *Project) appendChangeGrantMemberEvent(event *es_models.Event) error {
+	member := &ProjectGrantMember{}
+	err := member.getData(event)
+	if err != nil {
+		return err
+	}
+	for _, g := range p.Grants {
+		if g.GrantID == member.GrantID {
+			for i, m := range g.Members {
+				if m.UserID == member.UserID {
+					g.Members[i].getData(event)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (p *Project) appendRemoveGrantMemberEvent(event *es_models.Event) error {
+	member := &ProjectGrantMember{}
+	err := member.getData(event)
+	if err != nil {
+		return err
+	}
+	for _, g := range p.Grants {
+		if g.GrantID == member.GrantID {
+			for i, m := range g.Members {
+				if m.UserID == member.UserID {
+					g.Members[i] = g.Members[len(g.Members)-1]
+					g.Members[len(g.Members)-1] = nil
+					g.Members = g.Members[:len(g.Members)-1]
+				}
+			}
+		}
+	}
```

Comment:
the indentation is a bit high... and both loops just search for the member

i'd try to outsource at least part of the loop and the member deletion to improve readability

Reply:
could be used in eventstore.go as well